### ISSUE TITLE
Datapath program de-duplication

### ISSUE THREAD

**Issue by akshayknarayan** [ccp-project/portus#32]
@fcangialosi 's example:
```
impl<T: Ipc> CongAlg<T> for Bbr<T> {
  fn name() ...
  fn create() ...

  fn install_datapath_programs() {
    self.probe_rtt_program_id = portus::install_global_program(b"
        (def ... )
        (when ... )
    ")
    self.probe_bw_program_id = portus::install_global_program(b"
        (def ... )
        (when ... )
    ")
  }

  fn on_report() {
    if self.mode == probe_rtt {
      self.datapath.use_program(self.probe_rtt_program_id)
    }
    if self.mode == probe_bw {
      self.datapath.use_program(self.probe_bw_program_id)
    }
  }
}
```

We should de-duplicate programs, so that we can save datapath memory by avoiding storing programs per-flow. Flows share the program instructions, but obviously their state remains independent.

This subsumes #2 and #6

**Comment by akshayknarayan** [comment]
This may also help address https://github.com/ccp-project/ccp-kernel/issues/6

**Comment by akshayknarayan** [comment]
Closed by #38
